### PR TITLE
Feedback - teacher view of keep working

### PR DESCRIPTION
**Pull request by maureensturgeon**
Product spec: https://docs.google.com/document/d/1zYOOyTCG3FYpuuggQdRMOdJohI7IVngtbRjA6qbmN_A/edit#heading=h.sbw483gc8pm6
Tech spec: https://docs.google.com/document/d/1rPuJ8KBvGI0mYKfnWFgKxgBmiLPg3YJsIk2YnkNBF3Q/edit#

- Teacher feedback displays a new component `TeacherFeedbackKeepWorking` if the "keep working" experiment is enabled. This component can be in 3 possible states: indeterminate, checked and unchecked and is updated by the teacher. 
- Updates feedback status text to display "student updated at.." if the student has updated since feedback was left
- Teacher feedback serializer is removed and replaced by a TeacherFeedback `summarize` method which includes `student_last_updated` and `student_updated_since_feedback`

Screen Shot 2021-05-24 at 3 57 29 PM
Screen Shot 2021-05-24 at 4 00 25 PM

## Testing story
Added unit tests and tested following scenarios locally:
- [x] When teacher views feedback and is not part of the experiment, feedback looks as expected (with and without rubric)
- [x] When teacher views feedback and is part of experiment, looks as expected (for keep working, awaiting review, with and without rubric)
- [x] When a student views feedback, looks as expected (with and without rubric)

## Follow-up work
Move all `TeacherFeedback-` components into a TeacherFeedback directory.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/i18n/common/en_us.json
Comment (cforkish) [inline apps/i18n/common/en_us.json:173]: should "waiting" be capitalized?

File: apps/src/templates/instructions/CommentArea.jsx
Comment (cforkish) [inline apps/src/templates/instructions/CommentArea.jsx]: NIT: imo we should be consistent about only importing `React` and extending `React.Component`

File: apps/src/templates/instructions/CommentArea.jsx
Comment (cforkish) [inline apps/src/templates/instructions/CommentArea.jsx]: NIT: seeing how this is used, i think `isDisabled` or `isReadOnly` would be a clearer name (this name makes me need to figure out what is meant by "mode")

File: apps/src/templates/instructions/CommentArea.jsx
Comment (cforkish) [inline apps/src/templates/instructions/CommentArea.jsx:34]: could clean this up a bit: add a style `disabled: { backgroundColor: color.lightest_cyan }`, remove `textInputStudent` entirely, and just add `styles.disabled` when `disabledMode === true`

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (cforkish) [inline apps/src/templates/instructions/TeacherFeedback.jsx]: this could be simplified to `reviewState: latestFeedback?.review_state` (+ `|| null` if you don't want `undefined`)

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (cforkish) [inline apps/src/templates/instructions/TeacherFeedback.jsx]: NIT: from the name alone, it's not clear if this is a callback (e.g. `onFeedbackChanged`) or a property getter (e.g. `isFeedbackChanged` or `didFeedbackChange`). neither of those are very nice names though, so totally fine with leaving as-is.

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (cforkish) [inline apps/src/templates/instructions/TeacherFeedback.jsx]: storing this in `state` feels weird since it doesn't change. also getting it from the url feels hacky. i know this isn't code you wrote so it's not on you to clean it up if you don't want to, but would be nice to try to get this from redux and just have it in props.

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (cforkish) [inline apps/src/templates/instructions/TeacherFeedback.jsx:209]: why are we using a frontend experiment for this? won't we eventually want it to be a pilot that teachers can enroll in?

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (cforkish) [inline apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx]: NIT: imo enums should be capitalized (`ReviewStates`). also personally i think we should be consistent about using `utils.makeEnum` for this purpose.

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (cforkish) [inline apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx]: might as well validate this: `PropTypes.oneOf(Object.values(reviewStates))`

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (cforkish) [inline apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx]: NIT: delete the constructor and just set `checkbox = null` in its place

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (cforkish) [inline apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx]: i believe we had a conversation awhile back about consistently using `isXXX` for boolean properties and functions (along with `getXXX` and `renderXXX`). am i remembering that incorrectly?

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (cforkish) [inline apps/src/templates/instructions/TeacherFeedback.jsx:312]: looks like the `//Provided by Redux` section of `propTypes` needs cleanup

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (cforkish) [inline apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx]: NIT: `onCheckboxChange` seems to be a more common naming convention for event handlers in our code

**Review by cforkish — COMMENTED** [review]
overall this looks really great! i left a few questions and NITs, but nothing major.

one high-level comment though: this PR is pretty big, and any time you're introducing a brand new component, that's an easy opportunity to split out into a separate PR that we merge into the codebase before we start using it. no need to separate this one, but just wanted to call that out.

File: apps/i18n/common/en_us.json
Comment (maureensturgeon) [reply apps/i18n/common/en_us.json:173]: You can see an example of how it's used in the first screenshot, it's lowercase

Comment (cforkish) [reply apps/i18n/common/en_us.json:173]: 👍

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (maureensturgeon) [reply apps/src/templates/instructions/TeacherFeedback.jsx]: I'll set `this.studentId` instead of having it set on the state as a first step. I'm not sure we actually store the user that the teacher is viewing in redux

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (maureensturgeon) [reply apps/src/templates/instructions/TeacherFeedback.jsx:209]: This is not a frontend experiment, when a user is enrolled in a pilot, turns out the experiments the user is enrolled in is set on `window.appOptions.experiments` ( https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/helpers/levels_helper.rb#L314 ), and can be queried with experiments.js

Comment (cforkish) [reply apps/src/templates/instructions/TeacherFeedback.jsx:209]: oh interesting! i wonder if we should update the time spent pilot to do it that way at some point.

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (cforkish) [reply apps/src/templates/instructions/TeacherFeedback.jsx]: 👍

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (jamescodeorg) [inline apps/src/templates/instructions/TeacherFeedback.jsx:23]: I see a list of experiment names in experiments.js. Should this also go there?

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (jamescodeorg) [inline apps/src/templates/instructions/TeacherFeedback.jsx:52]: Mostly curious, why do we need `|| null` here?

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (jamescodeorg) [inline apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx]: Mostly curious, why are these snake-cased?

File: dashboard/app/models/teacher_feedback.rb
Comment (jamescodeorg) [inline dashboard/app/models/teacher_feedback.rb]: This is fine until we find a new pattern, but just want to point out that this line will query the user_levels table and couples this model tightly to the user model and how we store progress.  Rails makes this so (too!) easy.

One thing to watch out for -- if we ever retrieve a set of teacher feedbacks, this will likely result in a separate query to the user_levels table for each teacher feedback (a.k.a. the dreaded N+1 query problem).  This may or may not be ok depending on the call pattern.

File: apps/src/templates/instructions/TeacherFeedbackStatus.jsx
Comment (jamescodeorg) [inline apps/src/templates/instructions/TeacherFeedbackStatus.jsx]: I think this is the correct extension on the current implementation but after you merge this, I'd be curious if refactoring this component would improve readability.  I feel that hiding the various cases in functions kind of obscures the logic and wonder if putting all of the conditions in the main function and having each of them call a different render function would be a clearer way to factor the code.

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (jamescodeorg) [inline apps/src/templates/instructions/TeacherFeedback.jsx:226]: This div is getting kind of complicated.  As a future improvement, we may want to try to factor out a new component that represents the bottom of the teacher feedback panel.  (If it helps simplify things, I think we can propose moving the "last updated at" string.)

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (jamescodeorg) [inline apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx:16]: I think a header comment here (or maybe down near render?) explaining the two main states of this component ('normal' and 'awaiting teacher review') would be helpful.

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (jamescodeorg) [inline apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx]: Not sure if there's anything we can do about it, but having all of these styles here feels kind of complicated (and makes it a little hard to realize that there's only one line of text being rendered here).

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (jamescodeorg) [inline apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx:55]: Would it be simpler if we just stored and compared the initial state of the checkbox (unchecked, checked, indeterminate)?

**Review by jamescodeorg — COMMENTED** [review]
Phew, this was more complicated than I had anticipated!  I see that the indeterminate state and associated functionality was a part of this.

Thought for a future PR.  I think we may want to put the logic for "awaiting teacher review" -- and the rest of the state machine in the diagram in Req 4 -- on the back-end and reflect that in the review_state.  That seems like business logic that should belong on the back-end.

(Speaking of which, could you link the spec from the PR description?)

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (cforkish) [reply apps/src/templates/instructions/TeacherFeedback.jsx:23]: my thinking is that it should not go there. my initial thought when i saw this was that it was one of our frontend experiments like the ones listed in `experiments.js`, which are enabled in the browser by appending `?enableExperiments=myExperiment`, but apparently this value is passed down from the server through app options.

@maureensturgeon i do think it would be helpful to add a comment here since it tripped up both James and me.

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (maureensturgeon) [reply apps/src/templates/instructions/TeacherFeedback.jsx:52]: When performance level is checked and unchecked in the DOM the performance is set to null. So for consistency performance gets initialized to null instead of undefined. This isn't necessary though for the component to function, so I can remove it if you'd like.

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (maureensturgeon) [reply apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx]: The data is returned from endpoints that return feedback data as snake case (this was existing)

File: dashboard/app/models/teacher_feedback.rb
Comment (maureensturgeon) [reply dashboard/app/models/teacher_feedback.rb]: Yes! This is called in the teacher_feedback_serializer, which will result in multiple queries to user_levels if we return a set. We also make a call to `.teacher` in the serializer, so this seemed to follow the pattern we have in our codebase. I'm open to other options if we think this is going to cause issues.

File: apps/src/templates/instructions/TeacherFeedbackStatus.jsx
Comment (maureensturgeon) [reply apps/src/templates/instructions/TeacherFeedbackStatus.jsx]: I'll give it a go!

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (maureensturgeon) [reply apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx:55]: I gave this idea a try and decided that it didn't seem simplify things a whole lot, so I'm going to stick with the current logic

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (maureensturgeon) [reply apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx]: I've updated to use `&nbsp;`'s instead of spans around `-`

File: apps/src/templates/instructions/TeacherFeedbackStatus.jsx
Comment (maureensturgeon) [reply apps/src/templates/instructions/TeacherFeedbackStatus.jsx]: I've updated to have separate methods for each render, let me know what you think

File: dashboard/app/models/teacher_feedback.rb
Comment (maureensturgeon) [inline dashboard/app/models/teacher_feedback.rb:107]: This PR is already huge, so I didn't want to do the conversion to camelCase here

File: dashboard/app/models/teacher_feedback.rb
Comment (maureensturgeon) [inline dashboard/app/models/teacher_feedback.rb:98]: This was pulled in from the Teacher Feedback Serializer

File: apps/src/templates/instructions/TeacherFeedbackStatus.jsx
Comment (cforkish) [inline apps/src/templates/instructions/TeacherFeedbackStatus.jsx:29]: nice refactor! these individual render functions are much easier to follow.

File: dashboard/app/models/teacher_feedback.rb
Comment (cforkish) [reply dashboard/app/models/teacher_feedback.rb:107]: we should discuss what we want our conventions to be around this at some point. with few exceptions, our convention seems to be that we always return data from the server in snake_case and convert to camelCase on the frontend when (and if) we parse it. but i'm not convinced that's the best convention to stick to.

i would say though that since our current convention is to _not_ use camelCase on the backend, this is one of those TODO's that could end up sticking around for a long time, so might be better to track in a jira ticket rather than in an inline comment like this

**Review by cforkish — APPROVED** [review]

File: apps/src/templates/instructions/TeacherFeedback.jsx
Comment (jamescodeorg) [reply apps/src/templates/instructions/TeacherFeedback.jsx:52]: Ok leaving this, but maybe add a short comment?

File: dashboard/app/models/teacher_feedback.rb
Comment (jamescodeorg) [reply dashboard/app/models/teacher_feedback.rb:107]: If our current convention is to send snake_case from the back-end, I say we stick with that for now.  (Apologies if it was my comment that led to this, I didn't realize that was our current convention.)

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (jamescodeorg) [inline apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx:43]: Mostly curious, did you consider basing this logic on reviewState instead of the the checkbox state?

I think one of the things that I find a little bit confusing in this component is the mapping back and forth between the three conceptual states ('keep working', 'completed', 'awaiting review') and the three checkbox states ('checked', 'unchecked', 'indeterminate').  There is a 1:1 mapping between these, right?  If so, it might be nice to work primarily in one of these (the conceptual state?) and do the mapping only in display logic.

**Review by jamescodeorg — COMMENTED** [review]
One more question, but generally looks great!

File: dashboard/app/models/teacher_feedback.rb
Comment (maureensturgeon) [reply dashboard/app/models/teacher_feedback.rb:107]: Right now this is the method that we're using to format data and pass it to the frontend which is why I want to convert it to camelCase. But I agree with you, we need a pattern for how we go about serializing data for our frontend and for our casing in general. Other summarize methods that we have don't have a clear pattern, some are snake and some are camel case and some are both (https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/script.rb#L1413) 😬

Comment (maureensturgeon) [reply dashboard/app/models/teacher_feedback.rb:107]: I'll think about what the right move is here, maybe I'll remove the TODO and just leave snake case until we come up with a pattern.

File: apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx
Comment (maureensturgeon) [reply apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx:43]: It's not exactly a 1:1 because unchecked = completed or null depending on previous states

Comment (maureensturgeon) [reply apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx:43]: I've tried to refactor this several times and couldn't come up with something that was more straight forward. If you'd like to pair on a refactor, I'm super open to that!!

Comment (jamescodeorg) [reply apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx:43]: Ah, this is the key that I was missing -- there are *four* conceptual review states ('no value', 'keep working', 'completed', and 'awaiting review') and three checkbox states.  Thinking...

**Pull request merged**
